pkg/dynamic/client.go

```diff
@@ -94,3 +111,86 @@ func (c *knDynamicClient) ListSourcesTypes() (*unstructured.UnstructuredList, er
 func (c knDynamicClient) RawClient() dynamic.Interface {
 	return c.client
 }
+
+// ListSources returns list of available sources COs
+func (c *knDynamicClient) ListSources(f *SourceListFilters) (*unstructured.UnstructuredList, error) {
+	var sourceList unstructured.UnstructuredList
+	options := metav1.ListOptions{}
+	sourceTypes, err := c.ListSourcesTypes()
+	if err != nil {
+		return nil, err
+	}
+
+	namespace := c.Namespace()
+	// For each source type available, find out CO
+	for _, source := range sourceTypes.Items {
+		//  only find COs if this source type is given in filter
+		if f != nil && f.filters != nil {
+			// find source kind before hand to fail early
+			sourceKind, err := kindFromUnstructured(&source)
+			if err != nil {
+				return nil, err
+			}
+
+			// if this source is not given in filter flags continue
+			if !sliceContainsIgnoreCase(sourceKind, f.filters) {
+				continue
+			}
+		}
+
+		gvr, err := gvrFromUnstructured(&source)
+		if err != nil {
+			return nil, err
+		}
+
+		sList, err := c.client.Resource(gvr).Namespace(namespace).List(options)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(sList.Items) > 0 {
+			sourceList.Items = append(sourceList.Items, sList.Items...)
+		}
+		sourceList.SetGroupVersionKind(sList.GetObjectKind().GroupVersionKind())
```

Comment:
this looks wrong to me. The `sourceList` contains all elements from any type, so it looks for me that the last type just set the type of the sourceList (which is wrong as it can contain objects with different types)

Instead I would use for `sourceList` just a `[]*Unstructured` intstead of the list object (as we don't have a homogenous type)

Reply:
Yes, it sets the gvk for the last item processed. 

> Instead I would use for sourceList just a []*Unstructured intstead of the list object

Sorry, I didnt get this. 
Do you mean to say, don't set the GVK for sourceList?
(we needed GVK for machine readable flags for printers)

Reply:
I mean, don't use a `UnstructuredList` for the return value (except there is a specific reason for this, e.g. for the tabbed wrier), but a `[]*Unstructured` (so, only the `.items`), as the GVK for this list doesn't make sense for a potentially heterogenous list with different GVKs

Reply:
json/yaml/name printer requires the non `Empty` GVK for an object (List in this case), to print machinereadable output.

With following change: (i.e. setting `{Group: "", Version: "v1alpha1", Kind: "List"}` once for this list)
```
diff --git a/pkg/dynamic/client.go b/pkg/dynamic/client.go
index 1fc6e91a..71511343 100644
--- a/pkg/dynamic/client.go
+++ b/pkg/dynamic/client.go
@@ -151,8 +151,9 @@ func (c *knDynamicClient) ListSources(f *SourceListFilters) (*unstructured.Unstr
                if len(sList.Items) > 0 {
                        sourceList.Items = append(sourceList.Items, sList.Items...)
                }
-               sourceList.SetGroupVersionKind(sList.GetObjectKind().GroupVersionKind())
+
        }
+       sourceList.SetGroupVersionKind(schema.GroupVersionKind{Group: "", Version: "v1alpha1", Kind: "List"})
        return &sourceList, nil
 }
```
produce:

```
apiVersion: v1alpha1
items:
- apiVersion: sources.eventing.knative.dev/v1alpha1
  kind: ApiServerSource
  metadata:
    annotations:
      sources.eventing.knative.dev/creator: kube:admin
      sources.eventing.knative.dev/lastModifier: kube:admin
    creationTimestamp: "2020-02-25T16:58:36Z"
    generation: 1
    name: k8sevents
    namespace: default
    resourceVersion: "363874"
    selfLink: /apis/sources.eventing.knative.dev/v1alpha1/namespaces/default/apiserversources/k8sevents
    uid: 0e0bd443-5965-47e5-8ee5-76382208f5dc
  spec:
    mode: Ref
    resources:
    - apiVersion: v1
      controller: false
      controllerSelector:
        apiVersion: ""
        kind: ""
        name: ""
        uid: ""
      kind: Event
      labelSelector: {}
    serviceAccountName: events-sa
    sink:
      ref:
        apiVersion: serving.knative.dev/v1
        kind: Service
        name: eshow1
        namespace: default
  status:
    conditions:
    - lastTransitionTime: "2020-02-25T16:59:20Z"
      status: "True"
      type: Deployed
    - lastTransitionTime: "2020-02-25T16:59:34Z"
      message: apiserversources.sources.eventing.knative.dev are deprecated and will
        be removed in the future. Use apiserversources.sources.knative.dev instead.
      reason: ApiServerSourceDeprecated
      severity: Warning
      status: "True"
      type: Deprecated
    - lastTransitionTime: "2020-02-25T16:58:53Z"
      severity: Info
      status: "True"
      type: EventTypesProvided
    - lastTransitionTime: "2020-02-25T16:59:20Z"
      status: "True"
      type: Ready
    - lastTransitionTime: "2020-02-25T16:58:53Z"
      status: "True"
      type: SinkProvided
    - lastTransitionTime: "2020-02-25T16:58:53Z"
      status: "True"
      type: SufficientPermissions
    observedGeneration: 1
    sinkUri: http://eshow1.default.svc.cluster.local
kind: List
```
which is kind of similar to what `kubectl get sources -oyaml` produces (it also sets `Group` to "" and `apiVersion` to `v1`)
```
✗ kubectl get sources -oyaml  
apiVersion: v1
items:
- apiVersion: sources.eventing.knative.dev/v1alpha1
  kind: ApiServerSource
  metadata:
    annotations:
      sources.eventing.knative.dev/creator: kube:admin
      sources.eventing.knative.dev/lastModifier: kube:admin
    creationTimestamp: "2020-02-25T16:58:36Z"
    generation: 1
    name: k8sevents
    namespace: default
    resourceVersion: "363874"
    selfLink: /apis/sources.eventing.knative.dev/v1alpha1/namespaces/default/apiserversources/k8sevents
    uid: 0e0bd443-5965-47e5-8ee5-76382208f5dc
  spec:
    mode: Ref
    resources:
    - apiVersion: v1
      controller: false
      controllerSelector:
        apiVersion: ""
        kind: ""
        name: ""
        uid: ""
      kind: Event
      labelSelector: {}
    serviceAccountName: events-sa
    sink:
      ref:
        apiVersion: serving.knative.dev/v1
        kind: Service
        name: eshow1
        namespace: default
  status:
    conditions:
    - lastTransitionTime: "2020-02-25T16:59:20Z"
      status: "True"
      type: Deployed
    - lastTransitionTime: "2020-02-25T16:59:34Z"
      message: apiserversources.sources.eventing.knative.dev are deprecated and will
        be removed in the future. Use apiserversources.sources.knative.dev instead.
      reason: ApiServerSourceDeprecated
      severity: Warning
      status: "True"
      type: Deprecated
    - lastTransitionTime: "2020-02-25T16:58:53Z"
      severity: Info
      status: "True"
      type: EventTypesProvided
    - lastTransitionTime: "2020-02-25T16:59:20Z"
      status: "True"
      type: Ready
    - lastTransitionTime: "2020-02-25T16:58:53Z"
      status: "True"
      type: SinkProvided
    - lastTransitionTime: "2020-02-25T16:58:53Z"
      status: "True"
      type: SufficientPermissions
    observedGeneration: 1
    sinkUri: http://eshow1.default.svc.cluster.local
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

Reply:
Following up our discussion on slack: Added a plain sources's common fields holding struct values we want to print for 'kn source list'. The human readable output stays same, but see
```
./kn source list -oyaml
Items:
- Name: k8sevents
  Namespace: default
  Ready: "True"
  Resource: apiserversources.sources.eventing.knative.dev
  Sink: svc:eshow1
  SourceKind: ApiServerSource
kind: List
```